Move and consolidate address validation (#1886)
quicwg/base-drafts: draft-ietf-quic-transport.md

a different codepoint.

# Address Validation

Address validation is used by QUIC to avoid being used for a traffic
amplification attack.  In such an attack, a packet is sent to a server with
spoofed source address information that identifies a victim.  If a server
generates more or larger packets in response to that packet, the attacker can
use the server to send more data toward the victim than it would be able to send
on its own.

The primary defense against amplification attack is verifying that an endpoint
is able to receive packets at the transport address that it claims.  Address
validation is performed both during connection establishment (see

> {{validate-new}}) and during connection migration (see {{migrate-validate}}).

Author: In 7dc45966

## Address Validation During Connection Establishment {#validate-new}

During the initial handshake, QUIC requires that clients send UDP datagrams with
at least 1200 octets of payload until the server has completed address
validation. A server can thereby send more data to an unproven address without
increasing the amplification advantage gained by an attacker.

> A server eventually confirms that a client has received its messages when the

Reviewer: I'd drop eventually.

> first Handshake-level message is received. This might be insufficient, either

Reviewer: insufficient for what?

Reviewer: Or maybe this whole paragraph should be removed, because it's really not easy to read for me and I'm not sure what it's attempting to add.

because the server wishes to avoid the computational cost of completing the
handshake, or it might be that the size of the packets that are sent during the

> handshake is too large.  This is especially important for 0-RTT, where the

Reviewer: is -> are

token is repeated by the client in an Initial packet after it receives the Retry
packet.  In response to receiving a token in an Initial packet, a server can
either abort the connection or permit it to proceed.

A server can also use the Retry packet to defer the state and processing costs
of connection establishment.  By giving the client a different connection ID to
use, the connection might be routed to a different server instance, which have
additional capacity for new connections.

A flow showing the use of a Retry packet is shown in {{fig-retry}}.

~~~~
Client                                                  Server

Initial[0]: CRYPTO[CH] ->

                                                <- Retry+Token

Initial+Token[0]: CRYPTO[CH] ->

                                 Initial[0]: CRYPTO[SH] ACK[0]
                       Handshake[0]: CRYPTO[EE, CERT, CV, FIN]
                                 <- 1-RTT[0]: STREAM[1, "..."]
~~~~
{: #fig-retry title="Example Handshake Showing Retry"}

A connection MAY be accepted without address validation - or with only limited
validation - but a server SHOULD limit the data it sends toward an unvalidated

> address.  Successful completion of the cryptographic handshake implicitly

Reviewer: I think this last sentence is slightly incorrect.  Given it's covered above, maybe remove it?  Also, the limit towards an unvalidated address is covered above, so if there's anything important being added here, maybe move it above?

provides proof that the client has received packets from the server.

> ### Address Validation for Future Connections {#validation-future}

Reviewer: Technically this is still token based address validation, so the previous heading was a bit misleading.  Possibly it should be renamed Retry based address validation?

sending a Retry. If the validation succeeds, the server SHOULD then allow the
handshake to proceed.

Note:

: The rationale for treating the client as unvalidated rather than discarding
  the packet is that the client might have received the token in a previous
  connection using the NEW_TOKEN frame, and if the server has lost state, it
  might be unable to validate the token at all, leading to connection failure if
  the packet is discarded.  A server MAY encode tokens provided with NEW_TOKEN
  frames and Retry packets differently, and validate the latter more strictly.

In a stateless design, a server can use encrypted and authenticated tokens to
pass information to clients that the server can later recover and use to
validate a client address.  Tokens are not integrated into the cryptographic
handshake and so they are not authenticated.  For instance, a client might be
able to reuse a token.  To avoid attacks that exploit this property, a server
can limit its use of tokens to only the information needed validate client
addresses.

### Address Validation Token Integrity {#token-integrity}

There is no need for a single well-defined format for the token because the

> server that generates the token also consumes it.  A token could include

Reviewer: If the token includes this information, should we say something like.  If any information is included, it SHOULD be encrypted to avoid spoofing or leaking of information?

information about the claimed client address (IP and port), a timestamp, and any
other supplementary information the server will need to validate the token in
the future.

> An address validation token MUST be difficult to guess.  Including a large

Reviewer: This seems like the most critical requirement, so maybe it should be first?  Especially since the first and third paragraphs of this section flow together better than having this one in the middle.

{{QUIC-RECOVERY}}.  If the client has no data to retransmit and does not have
Handshake keys, it SHOULD send an Initial packet in a UDP datagram of at least
1200 octets.  If the client has Handshake keys, it SHOULD send a Handshake
packet.

A server might wish to validate the client address before starting the
handshake.  To send additional data prior to completing the cryptographic
handshake, the server then needs to validate that the client owns the address
that it claims.  QUIC therefore provides mechanisms for source address
validation during connection establishment.

Source addresses can be verified through an address validation token.  This
token is delivered during connection establishment with a Retry packet (see
{{validate-retry}}) or in a previous connection using the NEW_TOKEN frame (see
{{validate-future}}).

### Address Validation using Retry Packets {#validate-retry}

QUIC uses token-based address validation during connection establishment.  Any
time the server wishes to validate a client address, it provides the client with
a token.  As long as it is not possible for an attacker to generate a valid

> token for its address (see {{token-integrity}}) and the client is able to return

Reviewer 2: ```suggestion
token for its address (see {{token-integrity}}) and the client is able to return
token for its own address (see {{token-integrity}}) and the client is able to return
```

A server MAY provide clients with an address validation token during one
connection that can be used on a subsequent connection.  Address validation is
especially important with 0-RTT because a server potentially sends a significant
amount of data to a client in response to 0-RTT data.

The server uses the NEW_TOKEN frame {{frame-new-token}} to provide the client
with an address validation token that can be used to validate future
connections.  The client may then use this token to validate future connections
by including it in the Initial packet's header.  The client MUST NOT use the
token provided in a Retry for future connections.

Unlike the token that is created for a Retry packet, there might be some time
between when the token is created and when the token is subsequently used.
Thus, a resumption token SHOULD include an expiration time.  The server MAY
include either an explicit expiration time or an issued timestamp and
dynamically calculate the expiration time.  It is also unlikely that the client
port number is the same on two different connections; validating the port is
therefore unlikely to be successful.

> A resumption token SHOULD be easily distinguishable from tokens that are sent in

Reviewer 2: ```suggestion
A resumption token SHOULD be constructed to be easily distinguishable from tokens that
are sent in Retry packets as they are carried in the same field.
```

> Retry packets as they are carried in the same field.

Reviewer 2: ```suggestion
```

If the client has a token received in a NEW_TOKEN frame on a previous connection
to what it believes to be the same server, it can include that value in the
Token field of its Initial packet.

> A token allows a server to correlate activity between connections.

Reviewer 2: ```suggestion
A token allows a server to correlate activity between the connection where the token
was issued and any connection where it is used.
```

> Specifically, the connection where the token was issued, and any connection

Reviewer 2: ```suggestion
```

> where it is used.  Clients that want to break continuity of identity with a

Reviewer 2: ```suggestion
Clients that want to break continuity of identity with a
```

server MAY discard tokens provided using the NEW_TOKEN frame.  Tokens obtained
in Retry packets MUST NOT be discarded.

> A client SHOULD NOT reuse a token.  Reusing a token allows connections to be

Reviewer 2: ```suggestion
A client SHOULD NOT use a token more than once.  Reusing a token allows connections to be
```
